### PR TITLE
ECM-187: REST Service to return Enabled Planned Items w.r.t. a trackingId

### PR DESCRIPTION
**Pull request by marigianna**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -17,3 +17,5 @@ target/
 # IntelliJ
 .idea/
 *.iml
+/backend/.checkstyle
```
Comment (tobiasschaefer) [inline .gitignore]: please move these to a .gitignore file in the respective subdirectory

File: backend/build.sh [hunk 2]
```diff
@@ -5,7 +5,18 @@ START_TIME=$SECONDS
 
 echo -n "Java Version: " && java -version
 echo -n "Javac Version: " javac -version
-./mvnw install
```
Comment (tobiasschaefer) [inline backend/build.sh]: These changes should not be here. Please rebase correctly (or delete that commit)

File: backend/src/main/java/org/educama/shipment/api/datastructure/ShipmentPlanItemDS.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package org.educama.shipment.api.datastructure;
+
+/**
+ * @author msk
```
Comment (tobiasschaefer) [inline backend/src/main/java/org/educama/shipment/api/datastructure/ShipmentPlanItemDS.java]: please remove @author

File: backend/src/main/java/org/educama/shipment/boundary/impl/ShipmentBoundaryServiceImpl.java [hunk 4]
```diff
@@ -77,7 +77,8 @@ public ShipmentResource updateShipment(String trackingId, Shipment saveShipmentR
             shipment = shipmentRepository.saveAndFlush(shipment);
 
             if (completeShipmentOrderTask.isActive(trackingId) && completeShipmentOrderTask.canBeCompleted(trackingId)) {
-                completeShipmentOrderTask.complete(trackingId);
```
Comment (tobiasschaefer) [inline backend/src/main/java/org/educama/shipment/boundary/impl/ShipmentBoundaryServiceImpl.java]: the current requirement ist: if all data has been entered, then the shipment task is considered to be complete. Please revert.

File: backend/src/main/java/org/educama/shipment/boundary/impl/ShipmentTaskBoundaryServiceImpl.java [hunk 5]
```diff
@@ -43,4 +45,23 @@
...
+
+    @Override
+    public List<ShipmentPlanItemDS> findAllEnabled(String trackingId) {
+        List<ShipmentPlanItemDS> shipmentEnabledActivities = new ArrayList<ShipmentPlanItemDS>();
+        List<CaseExecution> enabledCaseExecutionsList = caseService.createCaseExecutionQuery().enabled().list();
+        for (CaseExecution caseExecution : enabledCaseExecutionsList) {
+            //returns the case instance
+            CaseInstance caseInstance = caseService.createCaseInstanceQuery().caseInstanceId(caseExecution.getCaseInstanceId()).singleResult();
+            if (caseInstance.getBusinessKey().equals(trackingId)) {
+            //returns the shipment for this caseinstance
```
Comment (tobiasschaefer) [inline backend/src/main/java/org/educama/shipment/boundary/impl/ShipmentTaskBoundaryServiceImpl.java]: fix indentation

File: backend/src/test/java/org/educama/common/api/RestApiDocumentation.java [hunk 6]
```diff
@@ -307,6 +343,9 @@ public void suggestCustomersTest() throws Exception {
                 );
     }
 
+
```
Comment (tobiasschaefer) [inline backend/src/test/java/org/educama/common/api/RestApiDocumentation.java]: are these new lines intentional?

**Comment by tobiasschaefer** [comment]
Please also fix you git history. It looks strange.

**Comment by marigianna** [comment]
Thanks Tobias, as its a bit messy now I will reorganize everything and create a new pull request on ECM-187.

**Comment by marigianna** [comment]
Hi, I hope it looks better now..please let me know.

File: backend/src/main/java/org/educama/shipment/api/ShipmentTaskController.java [hunk 7]
```diff
@@ -27,10 +34,30 @@
...
+    public ShipmentTaskListResource getActiveTasks() {
         List <ShipmentTaskDS> tasks = shipmentTaskBoundaryService.findAllActive();
         ShipmentTaskListResource taskList = new ShipmentTaskListResource().fromTaskCollection(tasks);
         return taskList;
     }
+    /**
+    * @param bases on the -by construction- precondition that the trackingId equals to the business key of the caseInstance
+    * @return a list of enabled activities for a particular shipment id
+    */
+   @RequestMapping(method = RequestMethod.GET, path = ShipmentTaskController.SHIPMENT_ENABLED_TASK_RESOURCE_PATH, produces = {MediaType.APPLICATION_JSON_VALUE})
```
Comment (tobiasschaefer) [inline backend/src/main/java/org/educama/shipment/api/ShipmentTaskController.java]: Please add an example request to the Postman collection, see Educama-Backend.postman_collection.json.
Please note: when exporting you should choose the format v1.

I haven't looked into the new REST interfaces in detail yet - the new requests in the Postman collection would make it easier :-)

The commit history and the other commits look good.

File: backend/src/main/java/org/educama/shipment/api/ShipmentTaskController.java [hunk 8]
```diff
@@ -1,24 +1,31 @@
...
  */
 @RestController
-@RequestMapping(path = ShipmentTaskController.SHIPMENTTASK_RESOURCE_PATH, produces = {MediaType.APPLICATION_JSON_VALUE})
 public class ShipmentTaskController {
 
-    public static final String SHIPMENTTASK_RESOURCE_PATH = "/educama/v1/tasks";
+    public static final String SHIPMENT_ACTIVE_TASK_RESOURCE_PATH = "/educama/v1/tasks";
+    public static final String SHIPMENT_ENABLED_TASK_RESOURCE_PATH = "/educama/v1/tasks/enabled";
+    public static final String SHIPMENT_COMPLETED_TASK_RESOURCE_PATH = "/educama/v1/tasks/completed";
+
```
Comment (larsduelfer) [inline backend/src/main/java/org/educama/shipment/api/ShipmentTaskController.java]: Hi @marigianna , I would like to discuss the API design here. Usually, I would expect to use query parameters like "..?state=enabled" or "...?state=completed. The next thing is, that they only make sense in the context of one shipment. The currently existing endpoint "/educama/v1/tasks" was just used for a generic task list (over all shipments) but we definely need an endpoint to get the tasks for one shipment. 

Another question then is how to get the active, enabled and completed once. 

So let us discuss this please.

File: backend/Educama-Backend.postman_collection.json [hunk 9]
```diff
@@ -1,189 +1,188 @@
 {
-	"id": "5e97a50a-323f-887a-8d9b-41c1a8b71ec6",
```
Comment (tobiasschaefer) [inline backend/Educama-Backend.postman_collection.json]: Please squash the last two commits regarding the postman collection into the commit which is related.

File: backend/src/main/java/org/educama/shipment/api/ShipmentTaskController.java (hunk 8 above)
Comment (marigianna) [inline backend/src/main/java/org/educama/shipment/api/ShipmentTaskController.java]: Hi @larsduelfer , thanks for the comments. 
Adding the parameters is ok, nevertheless, the enabled or completed should then call completely different functions..I guess this is not a problem...

I'm afraid I don't get the second part of the question...
Do you mean that we would like an extra service to return all tasks with respect to tracking Id but regardless of state?

**Comment by tobiasschaefer** [comment]
Your other PR https://github.com/Educama/Showcase/pull/126 has been merged into master. Please rebase against upstream/master and update this PR so that it only includes the REST API changes.

File: backend/src/main/java/org/educama/shipment/api/ShipmentTaskController.java (hunk 8 above)
Comment (marigianna) [inline backend/src/main/java/org/educama/shipment/api/ShipmentTaskController.java]: @larsduelfer  do we want to plan a ~30' meeting regarding this?

**Comment by tobiasschaefer** [comment]
As discussed today with HS Reutlingen (Jürgen): Let's close this PR. HS Reutlingen will create a new one.

**Pull request closed without merging**